File: Final_Project/Code/Project_4/Bai_toan_7/CSS_bt7.py
class Tree:
    def __init__(self, vertices):
        self.n = vertices
        self.adj = [[] for _ in range(vertices)]

    def add_edge(self, parent, child):
        self.adj[parent].append(child)

    def preorder(self, v):
        print(v, end=" ")
        for child in self.adj[v]:
            self.preorder(child)

    def postorder(self, v):
        for child in self.adj[v]:
            self.postorder(child)
        print(v, end=" ")

    def top_down(self, v):
        self.preorder(v)  # Top-down is equivalent to preorder

    def bottom_up(self, v):
        self.postorder(v)  # Bottom-up is equivalent to postorder

# Example usage
if __name__ == "__main__":
    t = Tree(5)
    t.add_edge(0, 1)
    t.add_edge(0, 2)
    t.add_edge(1, 3)
    t.add_edge(1, 4)
    print("Preorder:", end=" ")
    t.preorder(0)
    print()
    print("Postorder:", end=" ")
    t.postorder(0)
    print()
    print("Top-down:", end=" ")
    t.top_down(0)
    print()
    print("Bottom-up:", end=" ")
    t.bottom_up(0)
    print()

File: Final_Project/Code/Project_4/Bai_toan_7/CSS_bt7.cpp
#include <iostream>
#include <vector>
using namespace std;

class Tree {
public:
    vector<vector<int>> adj;
    int n;

    Tree(int vertices) : n(vertices) {
        adj.resize(n);
    }

    void add_edge(int parent, int child) {
        adj[parent].push_back(child);
    }

    void preorder(int v) {
        cout << v << " ";
        for (int child : adj[v]) {
            preorder(child);
        }
    }

    void postorder(int v) {
        for (int child : adj[v]) {
            postorder(child);
        }
        cout << v << " ";
    }

    void top_down(int v) {
        preorder(v); // Top-down is equivalent to preorder
    }

    void bottom_up(int v) {
        postorder(v); // Bottom-up is equivalent to postorder
    }
};

int main() {
    Tree t(5);
    t.add_edge(0, 1);
    t.add_edge(0, 2);
    t.add_edge(1, 3);
    t.add_edge(1, 4);
    cout << "Preorder: ";
    t.preorder(0);
    cout << endl;
    cout << "Postorder: ";
    t.postorder(0);
    cout << endl;
    cout << "Top-down: ";
    t.top_down(0);
    cout << endl;
    cout << "Bottom-up: ";
    t.bottom_up(0);
    cout << endl;
    return 0;
}